Add Watch to controller roles

kariya-mitsuru opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Starting from version 1.32, the client feature `WatchListClient` has been set to `true` in `kube-controller-manager`.
(commit 06a15c5cf96131faaf44f93f1be228a013ae5c0d)

As a result, when the `kube-controller-manager` executes the `List` method, it utilizes `Watch`. However, there are some existing controller roles that include `List` but do not include `Watch`. Therefore, when processes using these controller roles execute the `List` method, `Watch` is executed first, but due to permission errors, it falls back to `List`.

This PR adds `Watch` to the controller roles that include `List` but do not include `Watch`.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The following roles have had `Watch` added to them (prefixed with `system:controller:`):

- `cronjob-controller`
- `endpoint-controller`
- `endpointslice-controller`
- `endpointslicemirroring-controller`
- `horizontal-pod-autoscaler`
- `node-controller`
- `pod-garbage-collector`
- `storage-version-migrator-controller`
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

kariya-mitsuru committed:
Add Watch to controller roles

Starting from version 1.32, the client feature `WatchListClient` has been
set to `true` in `kube-controller-manager`.
(commit 06a15c5cf96131faaf44f93f1be228a013ae5c0d)

As a result, when the `kube-controller-manager` executes the `List` method,
it utilizes `Watch`. However, there are some existing controller roles that
include `List` but do not include `Watch`. Therefore, when processes using
these controller roles execute the `List` method, `Watch` is executed first,
but due to permission errors, it falls back to `List`.

This PR adds `Watch` to the controller roles that include `List` but do not
include `Watch`.

The affected roles are as follows (prefixed with `system:controller:`):

- `cronjob-controller`
- `endpoint-controller`
- `endpointslice-controller`
- `endpointslicemirroring-controller`
- `horizontal-pod-autoscaler`
- `node-controller`
- `pod-garbage-collector`
- `storage-version-migrator-controller`

Signed-off-by: Mitsuru Kariya <mitsuru.kariya@nttdata.com>

aojea commented:
/ok-to-test
/assign @p0lyn0mial @wojtek-t

aojea commented:
can we have some sort of linter or test that detect this?

aojea commented:
/assign @liggitt

kariya-mitsuru committed:
Fix Fixture Data

I apologize, the Fixture Data modifications were missed.

Signed-off-by: Mitsuru Kariya <mitsuru.kariya@nttdata.com>

kariya-mitsuru committed:
Add ControllerRoles Test

Added a test to check that if a controller role includes `List`, it also includes `Watch`.

Signed-off-by: Mitsuru Kariya <mitsuru.kariya@nttdata.com>

kariya-mitsuru commented:
> can we have some sort of linter or test that detect this?

Added a test to check that if a controller role includes `List`, it also includes `Watch`.

p0lyn0mial commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -95,7 +95,7 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 			rbacv1helpers.NewRule("update").Groups(batchGroup).Resources("cronjobs/status").RuleOrDie(),
 			rbacv1helpers.NewRule("update").Groups(batchGroup).Resources("cronjobs/finalizers").RuleOrDie(),
-			rbacv1helpers.NewRule("list", "delete").Groups(legacyGroup).Resources("pods").RuleOrDie(),
+			rbacv1helpers.NewRule("list", "watch", "delete").Groups(legacyGroup).Resources("pods").RuleOrDie(),
```
May I know how you discovered this issue? Was it just by scanning the logs?

kariya-mitsuru replied:
> May I know how you discovered this issue? Was it just by scanning the logs?

Yes, I noticed that `kube-controller-manager` is logging permission errors.

p0lyn0mial reviewed (commented):
We should also backport the fix.

/lgtm

liggitt commented:
> when processes using these controller roles execute the `List` method, `Watch` is executed first, but due to permission errors, it falls back to `List`.

I'm not 100% sure this is accurate... I thought WatchListClient *only* modified reflector behavior, which would already have had watch permissions... 

kube-controller-manager runs shared informers, not per-controller informers. Are we sure the individual list requests from these controllers are actually getting turned into watch requests and encountering permission errors?

/hold to verify this is needed / desired

liggitt commented:
> I'm not 100% sure this is accurate... I thought WatchListClient _only_ modified reflector behavior, which would already have had watch permissions...

checked the client and indeed, List was modified to attempt a Watch

we should add a unit test to this package to verify that every built-in role that grants list also grants watch

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy_test.go:
```diff
@@ -91,3 +92,15 @@ func TestControllerRoleLabel(t *testing.T) {
 	}
 }
+
+func TestControllerRoleVerbsConsistensy(t *testing.T) {
```
```suggestion
func TestControllerRoleVerbsConsistency(t *testing.T) {
```

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy_test.go:
```diff
@@ -91,3 +92,15 @@ func TestControllerRoleLabel(t *testing.T) {
 }
+
+func TestControllerRoleVerbsConsistensy(t *testing.T) {
+	roles := ControllerRoles()
```
should we also check NamespaceRoles(), NodeRules(), and ClusterRoles()?

kariya-mitsuru replied:
Added.

liggitt commented:
/hold cancel

kariya-mitsuru committed:
Fix typo

Co-authored-by: Jordan Liggitt <jordan@liggitt.net>

kariya-mitsuru committed:
Add Additional Tests

Added tests to check that if NodeRules, ClusterRoles, and NamespaceRoles
include `List`, it also include `Watch`.

Signed-off-by: Mitsuru Kariya <mitsuru.kariya@nttdata.com>

kariya-mitsuru commented:
/retest

wojtek-t commented:
this LGTM, but I want to ensure that Jordan doesn't have any follow-up comments before approving.

liggitt commented:
/retest

p0lyn0mial commented:
@kariya-mitsuru please don't forget to backport the fix to the previous version. Thanks!
